### PR TITLE
Run the SSP model simulation as separate process

### PR DESCRIPTION
**Pull request by adeas31**
Uses the zmq to read the simulation progress
Link with 3rdParty zmq lib
Save the SSP model before simulating

### CODE REVIEW

**Comment by adeas31** [comment]
Related to https://github.com/OpenModelica/OMSimulator/pull/931

File: OMEdit/OMEditLIB/OMS/OMSSimulationOutputWidget.cpp [hunk 1]
```diff
@@ -80,58 +88,176 @@ OMSSimulationOutputWidget::OMSSimulationOutputWidget(const QString &cref, QWidge
...
+    connect(mpSimulationProcess, SIGNAL(started()), SLOT(simulationProcessStarted()));
+    connect(mpSimulationProcess, SIGNAL(readyReadStandardOutput()), SLOT(readSimulationStandardOutput()));
+    connect(mpSimulationProcess, SIGNAL(readyReadStandardError()), SLOT(readSimulationStandardError()));
+  #if (QT_VERSION >= QT_VERSION_CHECK(5, 6, 0))
+    connect(mpSimulationProcess, SIGNAL(errorOccurred(QProcess::ProcessError)), SLOT(simulationProcessError(QProcess::ProcessError)));
+  #else
+    connect(mpSimulationProcess, SIGNAL(error(QProcess::ProcessError)), SLOT(simulationProcessError(QProcess::ProcessError)));
+  #endif
+    connect(mpSimulationProcess, SIGNAL(finished(int,QProcess::ExitStatus)), SLOT(simulationProcessFinished(int,QProcess::ExitStatus)));
+    QStringList args(QString("C:/OpenModelica/OMSimulator/src/OMSimulatorServer/OMSimulatorServer.py"));
```
Comment (adeas31) [inline OMEdit/OMEditLIB/OMS/OMSSimulationOutputWidget.cpp:126]: @lochel we need to copy this file to build directory. Right now the path is hard coded.

Comment (lochel) [reply OMEdit/OMEditLIB/OMS/OMSSimulationOutputWidget.cpp:126]: Can you do that?

**Comment by lochel** [comment]
Any hint how to launch the new simulation server? It seems that the simulation button still runs the "normal" simulation.

**Comment by AnHeuermann** [comment]
Oh and you changed the build system, so please add the CentOS and MINGW CI lables.

**Comment by lochel** [comment]
@AnHeuermann just try to rebase the pull request

**Comment by lochel** [comment]
Now I get the following error
```
clang++ -c -g -isystem . -isystem ../ -isystem /home/lochel/OpenModelica/build/include -isystem /home/lochel/OpenModelica/build/include/omplot -isystem /home/lochel/OpenModelica/build/include/omplot/qwt -isystem /home/lochel/OpenModelica/build/include/x86_64-linux-gnu/omc/antlr3 -isystem /home/lochel/OpenModelica/build/include/omc/scripting-API -isystem /home/lochel/OpenModelica/build/include/omc/c -isystem /home/lochel/OpenModelica/build/include/omc/c/util -isystem /home/lochel/OpenModelica/build/include/omc/fmil -isystem ../qjson/build/include -fPIC -Wall -W -D_REENTRANT -DQT_NO_DEBUG -DQT_WEBKITWIDGETS_LIB -DQT_WEBKIT_LIB -DQT_SVG_LIB -DQT_OPENGL_LIB -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_XMLPATTERNS_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_CONCURRENT_LIB -DQT_CORE_LIB -I. -I. -I.. -I../../build/include -I../../build/include/omplot -I../../build/include/omplot/qwt -I../../build/include/x86_64-linux-gnu/omc/antlr3 -I../../build/include/omc/scripting-API -I../../build/include/omc/c -I../../build/include/omc/c/util -I../../build/include/omc/fmil -I../qjson/build/include -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtWebKitWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtWebKit -isystem /usr/include/x86_64-linux-gnu/qt5/QtSvg -isystem /usr/include/x86_64-linux-gnu/qt5/QtOpenGL -isystem /usr/include/x86_64-linux-gnu/qt5/QtPrintSupport -isystem /usr/include/x86_64-linux-gnu/qt5/QtWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtXmlPatterns -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem /usr/include/x86_64-linux-gnu/qt5/QtXml -isystem /usr/include/x86_64-linux-gnu/qt5/QtConcurrent -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -Igeneratedfiles/moc -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o FMUSettingsDialog.o Animation/FMUSettingsDialog.cpp
OMS/OMSSimulationOutputWidget.cpp:42:10: fatal error: 'zmq.h' file not found
#include "zmq.h"
         ^~~~~~~
1 error generated.
make[3]: *** [Makefile:9568: OMSSimulationOutputWidget.o] Error 1
make[3]: *** Waiting for unfinished jobs....
```

**Comment by adeas31** [comment]
> Any hint how to launch the new simulation server? It seems that the simulation button still runs the "normal" simulation.

The same button launch the new simulation server. The old normal simulation is removed.

**Comment by adeas31** [comment]
> Now I get the following error
> 
> ```
> clang++ -c -g -isystem . -isystem ../ -isystem /home/lochel/OpenModelica/build/include -isystem /home/lochel/OpenModelica/build/include/omplot -isystem /home/lochel/OpenModelica/build/include/omplot/qwt -isystem /home/lochel/OpenModelica/build/include/x86_64-linux-gnu/omc/antlr3 -isystem /home/lochel/OpenModelica/build/include/omc/scripting-API -isystem /home/lochel/OpenModelica/build/include/omc/c -isystem /home/lochel/OpenModelica/build/include/omc/c/util -isystem /home/lochel/OpenModelica/build/include/omc/fmil -isystem ../qjson/build/include -fPIC -Wall -W -D_REENTRANT -DQT_NO_DEBUG -DQT_WEBKITWIDGETS_LIB -DQT_WEBKIT_LIB -DQT_SVG_LIB -DQT_OPENGL_LIB -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_XMLPATTERNS_LIB -DQT_NETWORK_LIB -DQT_XML_LIB -DQT_CONCURRENT_LIB -DQT_CORE_LIB -I. -I. -I.. -I../../build/include -I../../build/include/omplot -I../../build/include/omplot/qwt -I../../build/include/x86_64-linux-gnu/omc/antlr3 -I../../build/include/omc/scripting-API -I../../build/include/omc/c -I../../build/include/omc/c/util -I../../build/include/omc/fmil -I../qjson/build/include -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtWebKitWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtWebKit -isystem /usr/include/x86_64-linux-gnu/qt5/QtSvg -isystem /usr/include/x86_64-linux-gnu/qt5/QtOpenGL -isystem /usr/include/x86_64-linux-gnu/qt5/QtPrintSupport -isystem /usr/include/x86_64-linux-gnu/qt5/QtWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtXmlPatterns -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem /usr/include/x86_64-linux-gnu/qt5/QtXml -isystem /usr/include/x86_64-linux-gnu/qt5/QtConcurrent -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -Igeneratedfiles/moc -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o FMUSettingsDialog.o Animation/FMUSettingsDialog.cpp
> OMS/OMSSimulationOutputWidget.cpp:42:10: fatal error: 'zmq.h' file not found
> #include "zmq.h"
>          ^~~~~~~
> 1 error generated.
> make[3]: *** [Makefile:9568: OMSSimulationOutputWidget.o] Error 1
> make[3]: *** Waiting for unfinished jobs....
> ```

You need to compile OpenModelica first.

**Comment by AnHeuermann** [comment]
Ah, so there should be some dependencies to omedit in the Makefile.

**Comment by adeas31** [comment]
What do you mean?

**Comment by AnHeuermann** [comment]
I used `make -f Makefile.omdev.mingw all omedit -j`, but it didn't build omc before omedit.
I tried now to do `all` first and then compile `omedit` but it is still failing for me. Before 80fc4ad it was compiling.

I used ` git pull && git submodule update --force --init --recursive` to update this branch after 80fc4ad.

[build-omedit.log](https://github.com/OpenModelica/OpenModelica/files/5972086/build-omedit.log)

**Comment by adeas31** [comment]
Weird. `make -f Makefile.omdev.mingw omedit` should compile `omc` since we have this in the makefile,
```
omedit: omedit.skip
omedit.skip: omc omplot omsimulator omparser
```

Did you try to rebase the PR?

**Comment by AnHeuermann** [comment]
Yes, now it fails to compile because the 3rdParty submodule is at the wrong commit. I restarted the Jenkins job to see if it works on Windows or if it is just me who can't build it.

**Comment by adeas31** [comment]
Hmmm I guess I need to rebase and force push. But lets wait for the jenkins result.

**Comment by lochel** [comment]
It is running for more than 45 minutes now. I guess kill it and rebase the branch.

**Comment by AnHeuermann** [comment]
The Windows build is very slow, also there is a second one running on the Windows machines as well.

**Comment by lochel** [comment]
It doesn't make too much sense if it takes more then 45 minutes just to build omc. It didn't even start to run a single model... But I guess I am just too impatient.

**Comment by adeas31** [comment]
Yes the Windows build are annoyingly slow. I rebased and forced push. @AnHeuermann can you try now?

**Comment by AnHeuermann** [comment]
No, I did a hard reset and updated the submodules with `git submodule update --force --init --recursive` and recompiled.
But I still get
```
OMS/OMSProxy.cpp:1080:29: error: no matching function for call to 'oms_loadSnapshot'
  oms_status_enu_t status = oms_loadSnapshot(cref.toUtf8().constData(), snapshot.toUtf8().constData());
                            ^~~~~~~~~~~~~~~~
../../build/include\OMSimulator.h:125:33: note: candidate function not viable: requires 3 arguments, but 2 were provided
OMSAPI oms_status_enu_t OMSCALL oms_loadSnapshot(const char* cref, const char* snapshot, char** newCref);
````

**Comment by AnHeuermann** [comment]
Okay. I finally managed to compile OMEdit but now I have a new problem.

```
python D:/workspace/OpenModelica/build/share/OMSimulator/scripts/OMSimulatorServer.py --endpoint-pub=tcp://127.0.0.1:51498 --model=C:/OMDev/tools/msys/tmp/OpenModelica/OMEdit//Rectifier.ssp
python: can't open file 'D:/workspace/OpenModelica/build/share/OMSimulator/scripts/OMSimulatorServer.py': [Errno 2] No such file or directory
Simulation process failed. Exited with code 2.
```

**Comment by lochel** [comment]
Weird. That script should be installed into the correct location using OpenModelica/OMSimulator#944. What is your OMSimulator version?

**Comment by AnHeuermann** [comment]
>What is your OMSimulator version?

OMSimulator v2.1.1.post17-g064d498-mingw
Si it's the latest one.

But since Jenkins passed I'm sure I just did something wrong.

**Comment by lochel** [comment]
I don't know why, but I still cannot compile this pull request. I get the same error about zmq.h not found:

```bash
OMS/OMSSimulationOutputWidget.cpp:42:10: fatal error: 'zmq.h' file not found
#include "zmq.h"
         ^~~~~~~
1 error generated.
make[3]: *** [Makefile:9568: OMSSimulationOutputWidget.o] Error 1
```

**Comment by lochel** [comment]
This introduces a new dependency to zmq that wasn't installed on my linux: `libzmq3-dev`

Now it compiles at least and I can start testing.

**Review by lochel — APPROVED** [review]
I tested it on linux (wsl2) with a number of models and it seems to work really well 👍

**Pull request merged**